src/ir/term/mod.rs

```diff
@@ -721,6 +723,104 @@ impl Array {
         self.check_idx(idx);
         self.map.get(idx).unwrap_or(&*self.default).clone()
     }
+
+    /// Iter
+    pub fn into_iter(&self) -> std::collections::btree_map::IntoIter<Value, Value> {
```

Comment:
If you're going to implement the `into_iter` function, you should implement the `std::iter::IntoIterator` trait, not just add a member function.

Continuing my crusade against unnecessary cloning: you probably want `impl<'a> IntoIterator for &'a Array`, i.e., *not* a consuming iterator---just one that returns references. So I think what you really want is something like this (untested---see `impl<'a, K, V> IntoIterator for &'a BTreeMap<K, V>` in [the stdlib](https://doc.rust-lang.org/src/alloc/collections/btree/map.rs.html#1322)):

```rust
impl<'a> IntoIterator for &'a Array {
    type Item = (&'a Value, &'a Value);
    type IntoIter = std::collection::btree_map::Iter<'a, Value, Value>;
    fn into_iter(self) -> Self::IntoIter {
        self.map.iter()
    }
}
```

Reply:
A second, higher-level comment: it's extremely un-idiomatic for an `into_iter` function to take `&self` rather than `self`. This may seem slightly counterintuitive given what I've just said in my prior comment, but notice in my proposed `impl IntoIterator` that the argument to `fn into_iter` *is* `self`---it's just that we think of `self` as being a *value* of type `&'a Array`.

In contrast, what is currently implemented in the PR takes `&self` and then clones it to produce a by-value iterator, which is bad because it means that if someone did

```rust
    for (k, v) in some_array.into_iter() {
    }
```

they will have potentially caused a very expensive clone that they didn't expect (because idiomatically, into_iter never clones---it either consumes a value or it takes a reference and returns references).

Does this make sense?